Fixes #7369 - External user groups update on login

dLobatog opened the pull request:
This forces external user groups associated to the auth source of the user logging in to update, so that users get their external permissions updated on login.
Also there are a few unit tests that were large and I tried to tidy that up.

dLobatog commented:
- Remainder to update the security disclaimer in the LDAP docs after this is merged

ares commented on app/models/user.rb:
```diff
@@ -190,8 +190,9 @@ def self.try_to_login(login, password)
-          user.update_attributes(attrs.slice(:firstname, :lastname, :mail, :avatar_hash).delete_if { |k, v| v.blank? })
-        end if attrs.is_a? Hash
+          user.update_attributes(attrs.slice(:firstname, :lastname, :mail, :avatar_hash).delete_if { |k, v| v.blank? }) if attrs.is_a? Hash
+          user.auth_source.update_usergroups(login, password)
```
This will force login and password that is used during login but auth source may have explicit account to use for sync. In case where user does not have permissions he receives error "Insufficient Privileges to query groups data" upon login.

dLobatog commented:
@ares Thanks, I've updated it to call update_usergroups without using the user login and password, it uses the default which is the auth source account and account password.

dLobatog commented:
@ares I've made it to fallback to the user credentials if auth source credentials fail. However if these fail in the first place, it means the user cannot really be in there. It feels a little weird in my opinion though.

dLobatog commented:
@ares I've just rebased this

ares commented:
@eLobato it's not about fallback because of failing but supporting both use cases (first - you prefer to use credentials shared for all users stored on auth source, second - you use user's credentials directly, there's no shared credentials)

domcleal commented:
We don't support using a user's own credentials ($login) in an auth source in combination with LDAP linked user groups, so I wouldn't worry about that - just assume the shared credentials will work.

dLobatog commented:
@domcleal @ares I've changed it to make it use only the auth source 'shared' credentials, and also added a call in `try_to_auto_create_user` so external user groups are updated on user creation.

ares commented:
So the issue I struggled with was actually in ldap_fluff freeipa adapter. Here's the patch https://github.com/theforeman/ldap_fluff/pull/36 fixing it. Now it nearly works, there's one remaining issue though. We don't remove user groups correctly, I'll explain directly in code.

ares commented on app/models/auth_sources/auth_source_ldap.rb:
```diff
@@ -86,8 +86,8 @@ def ldap_con(login = nil, password = nil)
-  def update_usergroups(login, password)
-    ldap_con(login, password).group_list(login).each do |name|
+  def update_usergroups(login)
+    ldap_con(account, account_password).group_list(login).each do |name|
```
While this is not introduced by this PR, we have an issue here. Since we load only those groups that the user is assigned in LDAP we don't handle unassignment correctly. Suppose we have user in group A and B which are mapper to external groups of same name. Now admin removes user from external group B in LDAP. User logs into foreman and LDAP tells us the user is only in group A so we refresh only A. This results in user being assigned to B (it was not touched at all) until admin manually refresh the group B in foreman.

To solve it we should look at external groups (as we do now) plus all internal groups that user is assigned to. We should do union of groups found these way to avoid double refresh.

ares replied:
Plus few lines below, there's `rescue => error`, not sure how did that slip in and I know others don't like to mix refactoring into feature PRs but I think if you'll touch this area you could fix this as well. If it's too hard to name all possible exceptions, could you please `logger.debug error.backtrace.join("\n")`

dLobatog replied:
To the first comment: External user groups are currently validated for uniqueness in an auth source, so you cannot have two user groups pointing to the same external user group.

To the second: I'll add a few possibly known exceptions that I'll test, and keep the handler for StandardError just in case for other cases. The reason being that we don't want to lock out an user in case there's some error we didn't handle and on login external user groups cannot be updated :smile:

ares commented:
First: it's not about two user groups pointing to same external. Sorry, I didn't explain it clearly. Suppose there are A and B internal user groups and A and B in LDAP. A is mapped to A, B is mapped to B. User is associated to internal A and B. Admin removes his membership in B in LDAP. User logs in, so Foreman looks into LDAP and it finds out that user is assigned only to A, so it loads only internal A and run refresh just on it. It's 100% reproducible, so you can easily test that user remains in mapped internal group after you remove him from LDAP group when he login.

Second: I see, if you want to keep StandardError there I don't see a reason to name other exceptions. And we should make sure that we log the trace at least (using block).

dLobatog commented:
@ares Thanks for pointing out the mistake. I changed update_usergroups to look on the union of internal/external user groups when an user logs in so it can catch the bug you mentioned.

dLobatog commented:
@ares Rebased since it went out of date

dLobatog committed:
Fixes #7369 - External user groups update on login

ares commented:
I found another issue. External user group refresh fails if LDAP group is empty. For some reason we raise an exception in ldap fluff if we don't find any member for group (see https://github.com/theforeman/ldap_fluff/blob/master/lib/ldap_fluff/generic.rb#L47) so we should either fix ldap fluff or rescue this exception and make the user group empty. While this issue was not introduced by this PR, I'd be happy if we can address it here since it's needed for updating on login. But if you disagree I would be OK with merging this PR as it is.

Attaching the backtrace

```
RuntimeError: Group does not have any members
        from /opt/rh/ruby193/root/usr/share/gems/gems/ldap_fluff-0.3.2/lib/ldap_fluff/generic.rb:47:in `users_for_gid'
        from /opt/rh/ruby193/root/usr/share/gems/gems/ldap_fluff-0.3.2/lib/ldap_fluff/ldap_fluff.rb:35:in `user_list'
        from /usr/share/foreman/app/models/auth_sources/auth_source_ldap.rb:111:in `users_in_group'
        from /usr/share/foreman/app/models/external_usergroup.rb:37:in `users'
        from /opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/associations/collection_proxy.rb:89:in `map'
        from /opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/associations/collection_proxy.rb:89:in `method_missing'
        from /usr/share/foreman/app/models/external_usergroup.rb:20:in `refresh'
...cut...
```

dLobatog commented:
@ares Fixed, I rescue no particular exception here because ldap_fluff raises just a RuntimeError, it should be removed as soon as https://github.com/theforeman/ldap_fluff/pull/37 is merged though.

ares commented:
Thanks @eLobato, it works with patched ldap_fluff nicely. I'll merge when ldap_fluff patch gets in.

ares commented:
Merged as 01f8b024199f3bb2951ba1e7fe98959f74b9f3a4, thanks @eLobato!
